internal/datacoord/snapshot.go

```diff
@@ -0,0 +1,1214 @@
+│   ├── 00002-{uuid}.json          # Snapshot version 2 metadata (JSON format)
+│   └── ...
+│
+└── manifests/
+    ├── data-file-manifest-{uuid}.avro         # Manifest File containing segment details (Avro format)
+    └── ...
+
+Example paths:
+- s3://bucket/snapshots/12345/metadata/00001-a1b2c3d4.json
+- s3://bucket/snapshots/12345/manifests/data-file-manifest-i9j0k1l2.avro
+
+File Format Details:
+- metadata/*.json: JSON format containing snapshot metadata, collection schema, indexes, and manifest file paths array
+- manifests/data-file-manifest-*.avro: Avro format containing detailed segment information
+
+Access Pattern:
+1. Find the metadata JSON file by snapshot ID (iterate through metadata files)
+2. Read the metadata JSON file to get snapshot info and manifest file paths
+3. Read manifest Avro files to get segment file paths
+*/
+
+// SnapshotWriter is responsible for writing snapshots to S3.
+type SnapshotWriter struct {
+	chunkManager storage.ChunkManager
+}
+
+// NewSnapshotWriter creates a new writer instance.
+func NewSnapshotWriter(cm storage.ChunkManager) *SnapshotWriter {
+	return &SnapshotWriter{
+		chunkManager: cm,
+	}
+}
+
+// Save is the main entry point, executing the complete snapshot saving logic.
+// Returns the S3 path of the metadata file for the saved snapshot.
+func (w *SnapshotWriter) Save(ctx context.Context, snapshot *SnapshotData) (string, error) {
+	// Validate input parameters
+	if snapshot == nil {
+		return "", fmt.Errorf("snapshot cannot be nil")
+	}
+	if snapshot.SnapshotInfo == nil {
+		return "", fmt.Errorf("snapshot info cannot be nil")
+	}
+	collectionID := snapshot.SnapshotInfo.GetCollectionId()
+	if collectionID <= 0 {
+		return "", fmt.Errorf("invalid collection ID: %d", collectionID)
+	}
+	snapshotID := snapshot.SnapshotInfo.GetId()
+	if snapshotID <= 0 {
+		return "", fmt.Errorf("invalid snapshot ID: %d", snapshotID)
+	}
+	if snapshot.Collection == nil {
+		return "", fmt.Errorf("collection description cannot be nil")
+	}
+
+	basePath := path.Join(SnapshotRootPath, strconv.FormatInt(collectionID, 10))
+	writeUUID := uuid.New().String()
+
+	// Step 1: Write segment information to one or more Manifest files.
+	manifestFilePath, _, err := w.writeManifestFile(ctx, basePath, snapshotID, writeUUID, snapshot.Segments)
+	if err != nil {
+		return "", fmt.Errorf("failed to write manifest file: %w", err)
+	}
+	log.Info("Successfully wrote manifest file",
+		zap.String("manifestFilePath", manifestFilePath))
+
+	// Step 1.5: Collect StorageV2 manifest paths from segments.
+	storagev2Manifests := make([]*StorageV2SegmentManifest, 0)
+	for _, segment := range snapshot.Segments {
+		if segment.GetManifestPath() != "" {
+			storagev2Manifests = append(storagev2Manifests, &StorageV2SegmentManifest{
+				SegmentID: segment.GetSegmentId(),
+				Manifest:  segment.GetManifestPath(),
+			})
+		}
+	}
+
+	// Step 2: Create and write the core metadata.json file with manifest paths.
```

Comment:
Should we just write one file?

how to handle the orphan files if step 1 succeed but step2 failed?